### PR TITLE
DO entity mappings: initial contribution

### PR DESCRIPTION
**Pull request by stephan-merkli**
DO entity mappings allows to easily map the attributes of a DO entity to a form or any Java bean.

### CODE REVIEW

File: org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/mapping/AbstractDoEntityMapper.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+ * }
+ * </pre>
+ * <p>
+ * Usage Example within an {@link org.eclipse.scout.rt.client.ui.form.AbstractFormHandler}
+ * <ul>
+ * <li>execLoad: <code>BEANS.get(LoremFormMapper.class).fromDo(data, (LoremForm) form);</code>
+ * <li>execStore: <code>BEANS.get(LoremFormMapper.class).&#116;oDo((LoremForm)form, data);</code>
+ * </ul>
+ */
+public abstract class AbstractDoEntityMapper<DO_ENTITY extends IDoEntity, PEER> {
```
Comment (paolobazzi) [inline org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/mapping/AbstractDoEntityMapper.java:46]: not marked AppScoped on purpose?

File: org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/mapping/AbstractDoEntityMapper.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+  protected DoEntityMappings<DO_ENTITY, PEER> m_mappings = new DoEntityMappings<>();
+
+  public AbstractDoEntityMapper() {
+    initMappings(m_mappings);
```
Comment (paolobazzi) [inline org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/mapping/AbstractDoEntityMapper.java]: -> could call reset() here
-> m_mappings = new DoEntityMappings<>() above would not be necessary

File: org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/mapping/AbstractDoEntityMapper.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    extendMappings(m_mappings);
+  }
+
+  /**
+   * Extends the initialized mappings by mappings provided via {@link IDoEntityMapperExtension}.
+   */
+  @SuppressWarnings("unchecked")
+  protected void extendMappings(DoEntityMappings<DO_ENTITY, PEER> mappings) {
+    BEANS.all(IDoEntityMapperExtension.class).stream()
+        .filter(extension -> extension.getMapperClass().isInstance(AbstractDoEntityMapper.this)) // filter for those extensions that reference this mapper class as 3. generic parameter
```
Comment (paolobazzi) [inline org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/mapping/AbstractDoEntityMapper.java]: comment is not clear to me - which 3. parameter? why filter here?

File: org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/mapping/DoEntityMappings.java [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+import org.eclipse.scout.rt.platform.util.BooleanUtility;
+
+/**
+ * DO entity mappings are usually initialized via a subclass of {@link AbstractDoEntityMapper}.
+ */
+public class DoEntityMappings<DO_ENTITY extends IDoEntity, PEER> {
+
+  // Modifications to this list are not permitted once the instance of this class is created and initialized,
+  // otherwise ConcurrentModificationException can occur when to/fromDo is used concurrently with calls to the with methods.
+  protected final List<IDoEntityMapping<DO_ENTITY, PEER>> m_mappings = new ArrayList<>();
```
Comment (paolobazzi) [inline org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/mapping/DoEntityMappings.java:35]: AbstractDoEntityMapper could "seal" this mapping after init:

public AbstractDoEntityMapper() {
  initMappings(m_mappings);
  extendMappings(m_mappings);
	m_mappings.seal();  // block further modifications
}

**Review by paolobazzi — APPROVED** [review]
Looks good, see comments for some minor findings.

File: org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/mapping/AbstractDoEntityMapper.java (hunk 1 above)
Comment (stephan-merkli) [reply org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/mapping/AbstractDoEntityMapper.java:46]: Yes, subclass might decide for themself it that would like to create an application scopped mapper.

File: org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/mapping/DoEntityMappings.java (hunk 4 above)
Comment (stephan-merkli) [reply org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/mapping/DoEntityMappings.java:35]: No sealing for now, access via AbstractDoEntityMapper subclass is restricted, change of existing mappings is not supported.

**Pull request merged**